### PR TITLE
complete DB integration with AccountTree

### LINKED ISSUE #614: Implement DB Tree and DB Engine Logic
Currently the DB Tree and DB Engine are simply extensions of Memory Tree and Memory engine respective. Need to Implement separate logic for each to make it compatible with levelDB 

https://github.com/thehubbleproject/hubble-contracts/blob/master/ts/tree/dbTree.ts

https://github.com/thehubbleproject/hubble-contracts/blob/master/ts/client/database/databaseEngine.ts

### PR DESCRIPTION
**Pull request by kautukkundan**
closes #614

### CODE REVIEW

File: ts/tree/dbTree.ts [hunk 1]
```diff
@@ -1,17 +1,210 @@
...
+    constructor(depth: number, nodeType: string, hasher: Hasher) {
+        this.depth = depth;
+        this.hasher = hasher;
+        this.zeros = this.hasher.zeros(depth);
+        this.setSize = 2 ** this.depth;
+        this.nodeType = nodeType;
+    }
+
+    public async root(): Promise<Node> {
+        return await this.getNode(0, 0);
```
Comment (ChihChengLiang) [inline ts/tree/dbTree.ts:33]: We can add a `_root` variable to cache the root to make the `root` function stay sync. So that we don't have to change all over the codebase to use async function.

The `_root` should be updated in the `ascend` function.

Comment (kautukkundan) [reply ts/tree/dbTree.ts:33]: set the default value zeros[0] and create a function to set the root during restart

File: ts/tree/tree.ts [hunk 2]
```diff
@@ -29,3 +30,23 @@ export interface Tree {
     updateBatch(offset: number, leaves: Array<Node>): void;
     isZero(level: number, leafIndex: number): boolean;
 }
+
+export interface AsyncTree {
```
Comment (kautukkundan) [inline ts/tree/tree.ts:34]: change memory tree to use promises but use same interface for both memory and db tree

Comment (kautukkundan) [reply ts/tree/tree.ts:34]: keeping independent Async Tree to not mess with Memory Tree

File: package.json [hunk 3]
```diff
@@ -45,6 +45,7 @@
     "@types/subleveldown": "^4.1.0",
     "chai": "^4.3.4",
     "chai-as-promised": "^7.1.1",
+    "del": "^6.0.0",
```
Comment (kautukkundan) [inline package.json:48]: had to install this module because `rmdirAsync` from fs module wasn't compatible with node v10 for recursively deleting a folder.

File: ts/tree/dbTree.ts [hunk 4]
```diff
@@ -1,17 +1,218 @@
...
+import { ItemNode } from "./leaves/Node";
+import { AsyncTree, Data, Witness } from "./tree";
+
+export class DBTree implements AsyncTree {
+    public readonly zeros: Array<Node>;
+    public readonly depth: number;
+    public readonly setSize: number;
+    public readonly hasher: Hasher;
+    public readonly nodeType: string;
+    public _root: Node;
```
Comment (kautukkundan) [inline ts/tree/dbTree.ts:19]: using this variable as a cache for root which gets updated after every "ascend"

File: ts/tree/dbTree.ts [hunk 5]
```diff
@@ -1,17 +1,218 @@
...
+                len += 1;
+            }
+            for (let node = offset; node < offset + len; node += 2) {
+                await this.updateCouple(level, node);
+            }
+            offset >>= 1;
+            len >>= 1;
+        }
+
+        await this.updateRoot();
```
Comment (kautukkundan) [inline ts/tree/dbTree.ts:195]: updates the cached root

File: package.json (hunk 3 above)
Comment (jacque006) [reply package.json:48]: Can we remove this and use `rmdirAsync` once https://github.com/thehubbleproject/hubble-contracts/issues/567 is completed? If so I would add a comment to that issue to that effect.

File: test/client/integration.test.ts [hunk 6]
```diff
@@ -26,6 +27,14 @@ import * as mcl from "../../ts/mcl";
  * and reach a correct end state.
  */
 describe("Client Integration", function() {
+    before(async () => {
+        await del("./leveldb");
+    });
+
+    after(async () => {
+        await del("./leveldb");
+    });
```
Comment (jacque006) [inline test/client/integration.test.ts:36]:
```suggestion
    before(async function () {
        await del("./leveldb");
    });

    after(async function () {
        await del("./leveldb");
    });
```
See https://mochajs.org/#arrow-functions . Generally in our test suite it will not matter, but good to follow Mocha's recommendations. Arrow funcs are also generally fine as callbacks for `descibe` blocks.

File: test/client/integration.test.ts [hunk 7]
```diff
@@ -26,6 +27,14 @@ import * as mcl from "../../ts/mcl";
  * and reach a correct end state.
  */
 describe("Client Integration", function() {
+    before(async () => {
+        await del("./leveldb");
```
Comment (jacque006) [inline test/client/integration.test.ts:31]: We should make the leveldb path configurable in future work, out of scope for this PR.

File: ts/client/database/databaseEngine.ts [hunk 8]
```diff
@@ -1,12 +1,107 @@
...
+        return this.tree.root;
+    }
+
+    public async get(itemID: number): Promise<Item> {
+        this.checkSize(itemID);
+        try {
+            return (await this.leafFactory.fromDB(itemID)).item;
+        } catch (error) {
+            throw new Error(`Item not exists  itemID: ${itemID}`);
+        }
```
Comment (jacque006) [inline ts/client/database/databaseEngine.ts:40]: Is this `try/catch` needed, or will the underlying leveldb `get` in `leafFactory.fromDB` throw? If `get` can return an undefined value, this would be a better not found approach:
```suggestion
        const leaf  = await this.leafFactory.fromDB(itemID);
        if (!leaf) {
            throw new Error(`Item not exists  itemID: ${itemID}`);
        }
        return leaf.item;
```

File: ts/client/database/databaseEngine.ts [hunk 9]
```diff
@@ -1,12 +1,107 @@
...
+        return 0;
+    }
+
+    public revert(checkpoint: number = 0) {
+        // no-op
+    }
+
+    public async commit() {
+        // no-op
+    }
```
Comment (jacque006) [inline ts/client/database/databaseEngine.ts:75]: :+1:

File: ts/tree/dbTree.ts [hunk 10]
```diff
@@ -1,17 +1,218 @@
...
+    get root(): Node {
+        return this._root;
+    }
+
+    public async getNode(level: number, index: number): Promise<Node> {
+        try {
+            return await ItemNode.fromDB(this.nodeType, level, index);
+        } catch (e) {
+            return this.zeros[level];
+        }
```
Comment (jacque006) [inline ts/tree/dbTree.ts:43]: This could suppress other errors asides from the item not being found, can we do an undefined check on `ItemNode.fromDB` and throw instead?

File: ts/tree/leaves/Node.ts [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+): string => `${name}${depth}${index}`;
+
+export class ItemNode {
+    static async fromDB(
+        name: string,
+        depth: number,
+        index: number
+    ): Promise<Node> {
+        const key = getNodeKey(name, depth, index);
+        return await nodeDB.get(key);
```
Comment (jacque006) [inline ts/tree/leaves/Node.ts:17]: `await`ing is not needed in async chains when returning values, see this discussion https://github.com/thehubbleproject/hubble-contracts/pull/556#discussion_r618636732
```suggestion
        return nodeDB.get(key);
```

**Review by jacque006 — CHANGES_REQUESTED** [review]
Overall looks good. Some comments mainly around exception throwing/handling.

Would be great if we could get a test file that covers the async tree implementation for both memory and db, there's some starting logic in `stateEngine.test.ts` for the memory engine/tree. Would ensure we have edge cases and behavior properly covered, and that we can 1-1 swap tree/storage engine implementations. Think it would be fine to do that as follow up to this PR.

File: ts/tree/dbTree.ts [hunk 12]
```diff
@@ -1,17 +1,218 @@
...
+        return new DBTree(depth, nodeType, hasher || Hasher.new());
+    }
+
+    constructor(depth: number, nodeType: string, hasher: Hasher) {
+        this.depth = depth;
+        this.hasher = hasher;
+        this.zeros = this.hasher.zeros(depth);
+        this.setSize = 2 ** this.depth;
+        this.nodeType = nodeType;
+        this._root = this.zeros[0];
```
Comment (ChihChengLiang) [inline ts/tree/dbTree.ts:31]: I just comment to highlight issues here. We can keep it as is in this PR and fix the problem in the next PR.

This works when we have an empty tree. But we will have a problem once we have new leaves in the tree and have a non-empty tree root in DB.

When we restart the client and read the `root()` from the DBTree, we expect to read the non-empty tree root in DB. But here the cached root still returns the empty tree root.

We will need 2 steps. One reads the cache from DB and a step dumps the cache to DB.

File: package.json (hunk 3 above)
Comment (kautukkundan) [reply package.json:48]: Yeah sure we can

File: ts/client/database/databaseEngine.ts (hunk 8 above)
Comment (kautukkundan) [reply ts/client/database/databaseEngine.ts:40]: yeah, get throws an error if the key is not present in the db. 
`NotFoundError: Key not found in database`

File: test/client/integration.test.ts (hunk 6 above)
Comment (kautukkundan) [reply test/client/integration.test.ts:36]: ah, wasn't aware of this. will make changes.

File: ts/tree/leaves/Node.ts (hunk 11 above)
Comment (kautukkundan) [reply ts/tree/leaves/Node.ts:17]: gotcha, thanks for the detailed explanation 😄

File: ts/tree/dbTree.ts (hunk 12 above)
Comment (kautukkundan) [reply ts/tree/dbTree.ts:31]: absolutely! I had planned on discussing this on the call. We can come up with a `restoreFromDB` function which can check if a tree already exists and if yes, it can load the tree root. Was a bit confused though as to where we will be calling that.

File: ts/tree/dbTree.ts (hunk 10 above)
Comment (kautukkundan) [reply ts/tree/dbTree.ts:43]: check if key exists

**Review by jacque006 — APPROVED** [review]
